pandas/core/extensions.py

```diff
@@ -0,0 +1,216 @@
+"""Extend pandas with custom array types.
+"""
+import abc
+import typing as T
+
+import numpy as np
+
+
+class ExtensionDtype(metaclass=abc.ABCMeta):
+    """A custom data type for your array.
+    """
+    @property
+    @abc.abstractmethod
+    def type(self) -> type:
+        """Typically a metaclass inheriting from 'type' with no methods."""
+
+    @property
+    @abc.abstractmethod
+    def base(self) -> np.dtype:
+        # TODO
```

Comment:
It is not clear to me how this makes sense if the data is not backed by a NumPy array. I would be inclined to remove it from the interface if possible.

Perhaps it could be used by an optional `NumpyExtensionArray` base-class that pre-implements some methods if you are using a NumPy array as the backing store.